### PR TITLE
feat(aichat): add streaming support to ai chat and tutor via action cable and active job

### PR DESCRIPTION
**Pull request by drizco**
This pull request introduces streaming AI chat responses over ActionCable, allowing real-time delivery of assistant replies to users. It adds a new channel, job, and supporting helpers for streaming, and implements streaming support for both Gemini and OpenAI clients. A follow up pr for the front end will put this feature behind an experiment flag.

**Streaming infrastructure and channel integration:**

* Added `AichatChannel` for ActionCable-based AI chat streaming, handling subscription, request initiation, and error broadcasting. Integrates with the new streaming job and helper methods.
* Introduced `AichatRequestStreamJob` to process chat requests asynchronously, stream deltas as they arrive, handle toxicity checks, and broadcast completion or error events.

**AI client streaming support:**

* Implemented `stream_response` methods for both `AichatGeminiClient` and `AichatOpenaiResponsesClient`, enabling real-time SSE streaming from Gemini and OpenAI APIs, with parsing and error handling. 
* Added a reusable `process_sse_stream` method to `AichatAiClient` for parsing SSE events and yielding parsed JSON objects to callers.

**Helper and test improvements:**

* Enhanced `AichatAiHelper` with streaming response orchestration and broadcasting utilities for error and delta events.
* Added comprehensive tests for streaming, error, and edge cases in `AichatChannelTest`, and tested SSE parsing logic in `AichatAiClientTest`.

## Links

- Jira: https://codedotorg.atlassian.net/browse/LABS-1769

## PR Creation Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy impacts have been documented
- [ ] Security impacts have been documented
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Comment by snickell** [comment]
Associated frontend PR shows user visible behavior: https://github.com/code-dot-org/code-dot-org/pull/69942

File: dashboard/app/channels/aichat_channel.rb [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+      @stream_name = "aichat_stream:#{current_user.id}:#{stream_id}"
+      stream_from @stream_name
+    else
+      reject
+    end
+  end
+
+  # Expects payload similar to AichatRequestsController#start_chat_completion
+  # but streams the response over ActionCable via an ActiveJob
+  def request_completion(data)
```
Comment (edcodedotorg) [inline dashboard/app/channels/aichat_channel.rb:18]: As the above comment alludes, this whole file shares concerns with `dashboard/app/controllers/aichat_requests_controller.rb`.  Each of the helpers below: `build_request_attributes`, `handle_error`, `successful_stored_chat_messages`, and  `project_id_from_context` implement essentially what is implemented in `dashboard/app/controllers/aichat_requests_controller.rb` (with some differences) even if the implementation there  is not broken out into such helpers.  If these two paths (streaming and non) are going to live side by side for some time, I think it's important we do not have to make changes in two places whenever any of these concerns needs updated.

I propose we allow these two paths to share relevant code which could look like:

1. do an initial PR that focuses on the non-streaming case and pulls these concerns out into these helpers and put them in e.g. `aichat_ai_helper.rb`.  This first step would be focused on the non-streaming case so might have slightly different logic than what we have below.
2. do a follow on PR (or add to this PR) where we add an `isStreaming` param to these helpers (would now be in `aichat_ai_helper.rb`) to allow us to add any streaming specific logic.

@drizco Let me know your thoughts...

File: dashboard/app/helpers/aichat_gemini_client.rb [hunk 2]
```diff
@@ -1,10 +1,42 @@
 # This class implements a gemini backend for the generic AichatAiClient.
 class AichatGeminiClient < AichatAiClient
+  require 'net/http'
+
+  # Stream a response from Gemini, yielding each parsed chunk to the provided block.
+  def stream_response(config, request, context = [], &block)
+    AichatRubyTypes.assert_value_is_type(config, AichatAiClientTypes::AiConfig)
```
Comment (edcodedotorg) [inline dashboard/app/helpers/aichat_gemini_client.rb]: There is a lot of duplication in the Gemini and OpenAI specific `stream_response`.  

Screenshot 2026-01-06 at 3 36 53 PM

I propose `stream_response` is moved to the base class (in `aichat_ai_client.rb`) and any provider (Google vs OpenAI) specifics are moved into helpers which is the pattern generally used in these `aichat_*_client.rb` files.

Just glancing at the diff I think this could look like:

1. Add an `isStreaming` flag to `create_body` and let OpenAI's implementation add the `stream: true` if set.  Then the shared `stream_response` would just call `create_body` with this flag set.
2. Have OpenAI implement a `stream_url` that just returns `url` so `stream_response` can just have `uri = URI(stream_url)`
3. implement `http_open_timeout` and `http_read_timeout` methods
4. etc

File: dashboard/app/helpers/aichat_ai_client.rb [hunk 3]
```diff
@@ -52,6 +52,40 @@ def initialize(api_key, model, usage_reporter = nil)
...
+            end
+          end
+        end
+      end
+    end
+  end
+
   # The following methods MUST be Implemented By Derived class.
   # ------------------------------------------------------------
   # Raise NotImplementedError if these aren't implemented.
```
Comment (edcodedotorg) [inline dashboard/app/helpers/aichat_ai_client.rb:115]: The pattern here is if we add new required methods to the derived classes, we should be adding them as stubs with comments explaining what should be implemented and raising `NotImplementedError` through a call to `raise_not_implemented_error` (not sure we've been keeping up with that though).  This is really just documentation as to which methods have to be implemented by the derived classes as we're mimicking abstract classes.

File: dashboard/app/channels/aichat_channel.rb [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    payload = data.deep_stringify_keys
+
+    locale  = payload['locale'] || 'en'
+
+    request_attributes = build_request_attributes(payload)
+
+    request = AichatRequest.new(request_attributes)
+    request.save!
+
+    AichatRequestStreamJob.perform_later(
```
Comment (lfryemason) [inline dashboard/app/channels/aichat_channel.rb:28]: If the actioncable connection closes (e.g. user refreshes the page), do we need to be able to get all of the messages that already exist? 

We might already have that functionality somewhere else by retrieving some object in an http call, but wanted to call out that streaming from an actioncable does not get existing messages on connection.

**Review by lfryemason — COMMENTED** [review]
One note to think about, but otherwise LGTM from the actioncable side. I did not do a full review and only focused on the actioncable side.

File: dashboard/app/channels/aichat_channel.rb (hunk 4 above)
Comment (drizco) [reply dashboard/app/channels/aichat_channel.rb:28]: ah, great question! at the moment those messages would be lost because we currently rely on the client to send the completed chat events back to the server rather than letting the job persist them before returning them to the client. but I would like to change that, and this would be a good reason to make the change. more discussion in this pr https://github.com/code-dot-org/code-dot-org/pull/69950

File: dashboard/app/channels/aichat_channel.rb (hunk 1 above)
Comment (drizco) [reply dashboard/app/channels/aichat_channel.rb:18]: nice, good idea! I'll do a refactor of non streaming first and then update this pr

Comment (drizco) [reply dashboard/app/channels/aichat_channel.rb:18]: pr here! https://github.com/code-dot-org/code-dot-org/pull/70249

File: dashboard/app/helpers/aichat_gemini_client.rb (hunk 2 above)
Comment (drizco) [reply dashboard/app/helpers/aichat_gemini_client.rb]: @edcodedotorg pr has been updated to move stream_response to the base client. I'm not able to run the app locally unfortunately. I'll clean up the front end pr and mark it ready for review so you can test manually
